agent/rpc/peering/replication.go

```diff
@@ -282,8 +288,136 @@ func (s *Service) handleUpsertService(
 		}
 	}
 
-	// TODO(peering): cleanup and deregister existing data that is now missing safely somehow
+	//
+	// Now that the data received has been stored in the state store, the rest of this
+	// function is responsible for cleaning up data in the catalog that wasn't in the snapshot.
+	//
+
+	// nodeCheckTuple uniquely identifies a node check in the catalog.
+	// The partition is not needed because we are only operating on one partition's catalog.
+	type nodeCheckTuple struct {
+		checkID types.CheckID
+		node    string
+	}
+
+	var (
+		// unusedNodes tracks node names that were not present in the latest response.
+		// Missing nodes are not assumed to be deleted because there may be other service names
+		// registered on them.
+		// Inside we also track a map of node checks associated with the node.
+		unusedNodes = make(map[types.NodeID]struct{})
+
+		// deletedNodeChecks tracks node checks that were not present in the latest response.
+		// A single node check will be attached to all service instances of a node, so this
+		// deduplication prevents issuing multiple deregistrations for a single check.
+		deletedNodeChecks = make(map[nodeCheckTuple]struct{})
+	)
+	for _, csn := range storedInstances {
+		if _, ok := snap.Nodes[csn.Node.ID]; !ok {
+			unusedNodes[csn.Node.ID] = struct{}{}
+
+			// Since the node is not in the snapshot we can know the associated service
+			// instance is not in the snapshot either, since a service instance can't
+			// exist without a node.
+			// This will also delete all service checks.
+			err := s.Backend.Apply().CatalogDeregister(&structs.DeregisterRequest{
+				Node:           csn.Node.Node,
+				ServiceID:      csn.Service.ID,
+				EnterpriseMeta: csn.Service.EnterpriseMeta,
+				PeerName:       peerName,
+			})
+			if err != nil {
+				return fmt.Errorf("failed to deregister service: %w", err)
```

Comment:
Do we want to fail fast here vs using `merr` and grabbing all the raft apply errors? Wondering if it is meaningful to output node/service/check IDs in the error

Reply:
Hmmm I'm a bit hesitant to use `merr` for raft apply errors because the error could get really unwieldy. For example, if leadership is lost and all raft applies fail because of it there could be thousands of errors

Thoughts @rboyer ?

Reply:
Almost all times a raft apply will work. If it doesn't work it almost certainly is going to be due to leadership loss so we'd want fast failure at the first raft error, not collection of errors and display at the end.